### PR TITLE
fix: image loss on refresh and stop refreshes when done computing

### PR DESCRIPTION
**Pull request by tszirr**
# Bugfix
* On output buffer refresh, the output buffer was cleared, but images were not re-rendered (missing matching canvas.clear)
* If execution time is shown, the output window never stopped refreshing even after computations were done (missing check of starting_status)

### CODE REVIEW

File: rplugin/python3/molten/outputbuffer.py [hunk 1]
```diff
@@ -292,6 +292,7 @@ def show_floating_win(self, anchor: Position) -> None:
 
         # Clear buffer:
         self.nvim.funcs.deletebufline(self.display_buf.number, 1, "$")
+        self.canvas.clear()
```
Comment (benlubas) [inline rplugin/python3/molten/outputbuffer.py:295]: this line causes images that are part of virtual text output to disappear when I open the output in a floating window.

**Review by benlubas — COMMENTED** [review]
Could you provide better descriptions of the bug that you're fixing by adding the `canvas.clear` line?

The other fix is good, and I would merge that fix. 
In the future, please limit one bug fix per PR if they're unrelated, that way I don't have to stall both fixes b/c of issues or questions with one of them!

Also, I'd appreciate issues that describe the bug and how to reproduce it so that I can better test out bug fix PRs like this.

File: rplugin/python3/molten/outputbuffer.py (hunk 1 above)
Comment (tszirr) [reply rplugin/python3/molten/outputbuffer.py:295]: Virtual output was always broken for me, so I assumed this was by design/not a priority. To have well-defined behavior between virtual output and floating windows, I assume a separate canvas would be necessary for each?

**Comment by tszirr** [comment]
The bug that is addressed is that images are not redrawn after the text buffer is cleared, therefore there are no more extmarks or virtual lines in the updated output buffer. On second checking, it indeed appears that the ImageCanvas.clear function is not doing everything that would be needed to fix this. I can't seem to find a call site where it is referenced, what is it currently used for?

**Comment by tszirr** [comment]
Looking at ImageCanvas, it seems ~the actual bug this was fixing was~ that ImageCanvas.images is currently never filled, ~therefore it just keeps adding more images instead of reusing existing ones~ therefore it relies on image.nvim to resolve the redundancy. Thus, indirectly the fix is achieving the right thing, i.e. images are re-rendered correctly every time they are placed because the canvas was cleared.

Given that, it would of course be better to clean up the ImageCanvas behavior and achieve this more directly. Still, I am unsure how overlapping virtual output would work with images. I assume no image virtual lines can be added for padding in this case? Would that require a separate, non-inline image embedding branch?

File: rplugin/python3/molten/outputbuffer.py (hunk 1 above)
Comment (benlubas) [reply rplugin/python3/molten/outputbuffer.py:295]: Currently I'm able to show images as virtual text, open a popup, and see the image there as well. The image doesn't disappear when I close the output (although sometimes image.nvim bugs and doesn't render the image correctly until I scroll).

You can open an issue for virtual text if it's not working for you.

**Comment by benlubas** [comment]
> ImageCanvas.images is currently never filled

Oh that's true, leftover from when I switched to use image.nvim I guess, and I never really noticed. Similar story for the `canvas.clear` method. I use `remove_image` now to remove a single image at a time. I will clean those up.

> The bug that is addressed is that images are not redrawn after the text buffer is cleared, therefore there are no more extmarks or virtual lines in the updated output buffer

I'm still a little confused by what this means, could you give a brief reproduction, some code to run, and the molten commands you run/actions you take to observe the bug?

**Comment by tszirr** [comment]
Currently, from what I can gather, the output text buffer is cleared and refreshed many times, but images are only correctly placed in this buffer the first time they show up via their `render` method, which does not detect changes in the its extmarks after the buffer has been cleared, since the placement is not checked in the lazy update:
```
      local has_up_to_date_extmark = previous_extmark and previous_extmark.height == height
```
This is most noticeable in the fact that virtual lines are lost on update:
image

The clear call added in this PR fixed that, because it invalidates the extmark so that it is re-placed when the output buffer is re-assembled every time, re-creating the correct virtual lines every time:
image

As to the question how to consistently solve correct image output in the alternative output method/"virtual output" with the same images, I am unsure if this is possible with the current infrastructure, given that the padding method using virtual lines already seems incompatible with "virtual output" by design? It sounds like a deeper rework would be necessary to correctly support images there.

**Comment by benlubas** [comment]
I see what you mean. I've never used this plugin to render multiple images so I'm not surprised that it doesn't work well. to be honest, I don't have the motivation to work on this as much as I used to. I don't have to work with the Jupyter kernel at all now that I'm out of school and am not a data scientist.

I will be here to review and accept PRs to fix these things but I ask that you:
- open an issue first with clear instructions for how to reproduce each bug 
- Only fix one thing in one PR. Open a second to fix other issues
- Test your changes thoroughly with floating window and virtual text

I will close this for now, and you can reopen a PR with just the fix for:

> If execution time is shown, the output window never stopped refreshing even after computations were done (missing check of starting_status)

Then if you'd like, take a shot at fixing the other issue in a way that's compatible with virtual text that would be awesome

**Pull request closed without merging**